### PR TITLE
WIP - convert public React core properties to strings so uglify/gcc can mangle

### PR DESCRIPTION
**Pull request by trueadm**
Ideally, we want to be able to leverage stronger minifcations/optimizations in our tooling so we can provide even smaller bundles. Tools such as Uglify and Google Closure Compiler can do this when set to mangle internal properties. However, this can break things when properties are expected to be consumed in the wild/public or by contract from another module. 

The simplest way to deal with this limitation is to manually specify that the property shouldn't be altered when it's a string literal. Whitelisting is also an option, however that has some limitations and can result in properties that aren't public not being mangled simply because of them being a common name.

This PR is a WIP.

### CODE REVIEW

**Comment by sebmarkbage** [comment]
You're missing the base class and the updater. That might set you back a bit.

https://github.com/facebook/react/blob/master/src/isomorphic/modern/class/ReactBaseClasses.js#L85-L86

https://github.com/facebook/react/blob/master/src/isomorphic/modern/class/ReactNoopUpdateQueue.js

**Comment by spicyj** [comment]
Does this mean we need to change every single native method call? Ex: .map(), .indexOf(), etc.

**Comment by trueadm** [comment]
@spicyj No, we shouldn't have to. They should all be natively supported. Well, they definitely are in gcc.

File: src/shared/utils/ReactElementSymbol.js [hunk 1]
```diff
@@ -14,9 +14,9 @@
 
 // The Symbol used to tag the ReactElement type. If there is no native Symbol
 // nor polyfill, then a plain number is used for performance.
-var REACT_ELEMENT_TYPE = (typeof Symbol === 'function' &&
-  Symbol.for &&
-  Symbol.for('react.element')) ||
+const SymbolFor = Symbol.for;
```
Comment (simonkberg) [inline src/shared/utils/ReactElementSymbol.js]: This will throw in environments without Symbol

**Comment by gaearon** [comment]
Tracking in https://github.com/facebook/react/issues/11092.

**Pull request closed without merging**